File: server/entities/docindex.py
import csv
import json
import time
from typing import Dict, Optional, Tuple, List

class DocumentIndex:
    def __init__(self):
        self.csv_path = "server/data/test_100k.csv"
        self.index_path = "server/data/document_index.json"
        # Load the index once during initialization
        try:
            with open(self.index_path, 'r') as index_file:
                self.index = json.load(index_file)
        except FileNotFoundError:
            self.index = {}
        # Keep the CSV file handle as an instance variable
        self.csv_file = None

    def __enter__(self):
        """Context manager entry - opens the CSV file"""
        start = time.time()
        self.csv_file = open(self.csv_path, 'r', encoding='utf-8')
        end = time.time()

        print("mmaz kelite", end-start)
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit - closes the CSV file"""
        if self.csv_file:
            self.csv_file.close()
            self.csv_file = None

    def read_next_record(self, csv_file) -> Tuple[Optional[str], int]:
        """Read next CSV record, returning doc_id and start position."""
        start_pos = csv_file.tell()
        
        chunks = []
        while True:
            chunk = csv_file.readline()
            if not chunk:
                return None, 0
                
            chunks.append(chunk)
            try:
                record = next(csv.reader([''.join(chunks)]))
                return record[0], start_pos
            except csv.Error:
                continue
    
    def build_index(self) -> None:
        with open(self.index_path, 'w') as index_file:
            index_file.write('{\n')
            
            with open(self.csv_path, 'r', encoding='utf-8') as csv_file:
                csv_file.readline()  # Skip header
                first = True
                
                while True:
                    doc_id, pos = self.read_next_record(csv_file)
                    if doc_id is None:
                        break
                        
                    if not first:
                        index_file.write(',\n')
                    else:
                        first = False
                        
                    index_file.write(f'"{doc_id}": {pos}')
                    
            index_file.write('\n}')

    def get_documents(self, doc_ids: List[str]) -> List[Optional[Dict]]:
        """Get multiple documents by their IDs efficiently."""
        if not self.csv_file:
            raise RuntimeError("DocumentIndex must be used as a context manager")

        results = []
        for doc_id in doc_ids:
            if doc_id not in self.index:
                results.append(None)
                continue

            self.csv_file.seek(self.index[doc_id])
            content = []
            
            while True:
                chunk = self.csv_file.readline()
                content.append(chunk)
                try:
                    reader = csv.DictReader([''.join(content)],
                                          fieldnames=['id', 'title', 'keywords', 'venue', 
                                                    'year', 'n_citation', 'url', 'abstract',
                                                    'authors', 'doc_type', 'references'])
                    results.append(next(reader))
                    break
                except csv.Error:
                    continue

        return results

    def get_document(self, doc_id: str) -> Optional[Dict]:
        """Get a single document by ID"""
        if not self.csv_file:
            raise RuntimeError("DocumentIndex must be used as a context manager")

        if doc_id not in self.index:
            return None
            
        self.csv_file.seek(self.index[doc_id])
        content = []
        
        while True:
            chunk = self.csv_file.readline()
            content.append(chunk)
            try:
                reader = csv.DictReader([''.join(content)],
                                      fieldnames=['id', 'title', 'keywords', 'venue', 
                                                'year', 'n_citation', 'url', 'abstract',
                                                'authors', 'doc_type', 'references'])
                return next(reader)
            except csv.Error:
                continue
        
if __name__ == "__main__":
    with DocumentIndex() as documentIndex:
        documentIndex.build_index()
        # doc = documentIndex.get_document('0000a1fd')
        # print(doc)

File: server/testing/docindex.go
package main

import (
	"encoding/csv"
	"encoding/json"
	"io"
	"os"
)

type Document struct {
	ID         string `json:"id"`
	Title      string `json:"title"`
	Keywords   string `json:"keywords"`
	Venue      string `json:"venue"`
	Year       string `json:"year"`
	NCitation  string `json:"n_citation"`
	URL        string `json:"url"`
	Abstract   string `json:"abstract"`
	Authors    string `json:"authors"`
	DocType    string `json:"doc_type"`
	References string `json:"references"`
}

// loadIndex loads document offsets from JSON file
func LoadIndex() (map[string]int64, error) {
	data, err := os.ReadFile("../data/document_index.json")
	if err != nil {
		return nil, err
	}

	var index map[string]int64
	if err := json.Unmarshal(data, &index); err != nil {
		return nil, err
	}

	return index, nil
}

// getDocuments retrieves documents by their IDs
func GetDocuments(csvPath string, docIDs []string, index map[string]int64) ([]*Document, error) {
	file, err := os.Open(csvPath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	results := make([]*Document, len(docIDs))

	for i, docID := range docIDs {
		offset, exists := index[docID]
		if !exists {
			results[i] = nil
			continue
		}

		if _, err := file.Seek(offset, 0); err != nil {
			return nil, err
		}

		reader := csv.NewReader(file)
		record, err := reader.Read()
		if err != nil {
			if err == io.EOF {
				results[i] = nil
				continue
			}
			return nil, err
		}

		results[i] = &Document{
			ID:         record[0],
			Title:      record[1],
			Keywords:   record[2],
			Venue:      record[3],
			Year:       record[4],
			NCitation:  record[5],
			URL:        record[6],
			Abstract:   record[7],
			Authors:    record[8],
			DocType:    record[9],
			References: record[10],
		}
	}

	return results, nil
}

func main1() {
	// indexPath := "../data/document_index.json"
	csvPath := "../data/test_100k.csv"
	docIDs := []string{"1", "2", "3", "4", "5"}
	documentIndex, err := LoadIndex()

	docs, err := GetDocuments(csvPath, docIDs, documentIndex)
	if err != nil {
		panic(err)
	}

	for _, doc := range docs {
		if doc != nil {
			println(doc.ID)
		}
	}
}
